### ISSUE TITLE
request: Automatically create GitHub releases

### ISSUE THREAD

**Issue by Almighty-Alpaca** [invertase/melos#277]
### Is there an existing feature request for this?

- [X] I have searched the existing issues.

### Description

I propose adding a way to automatically create releases on GitHub in addition to the normal git tags. There are two way this could be achieved: 

- Authenticating with the official GitHub API

- Generating a link to a prefilled form for the user (I'd prefer this option). GitHub supports prefilling the form for creating a release using query parameters:
```url
https://github.com/invertase/melos/releases/new
	?tag=melos-v1.0.0
	&title=meslos%20-%20v1.0.0
	&body=Changelog%20and%20breaking%20information
	&prerelease=false
```

The process could triggered either
- manually with a command line flag: `melos version --github`
- it's own command: `melos github`
- automatically when the repository field in the `pubspec.yaml` is a GitHub repository
- a melos.yaml config entry: 
```yaml
command:
    version:
        github: true
```

It would also be possible to chose a more generic name to open up the feature for other platforms such as GitLab, however I'm not sure if GitLab supports prefilling the form in any way.

### Reasoning

I really love how much work melos saves with the `version` and `publish` commands. Creating releases on GitHub fits really well into the work melos already automates and is quite tiring to do manually every time.

**Comment by donheshanthaka** [comment]
Hey @Salakar  i have some doubts regarding this issue. Do you require a GitHub actions workflow to be implemented on this repo. Therefore, let's say when a commit is being made using a certain tag ex: `v`, the release will start automatically.

**Comment by Almighty-Alpaca** [comment]
I have recently gone ahead and implemented this in two stages:
- A [simple version](https://github.com/invertase/melos/compare/main...Almighty-Alpaca:melos:feat/github-release-url) that generates links to a prefilled form for GitHub releases
- A [more advanced version](https://github.com/invertase/melos/compare/main...Almighty-Alpaca:melos:feat/github-automatic-release) that uses the GitHub API and the `GITHUB_TOKEN` environment variable present in e.g. GitHub Actions to generate releases automatically.

The latter is currently untested as I noticed a few things while implementing this:
- Automatically generating a release doesn't make sense before pushing the release commit and tags. Doing that automatically is currently marked with [TODO](https://github.com/invertase/melos/blob/main/packages/melos/lib/src/commands/version.dart#L333) in the code
- Implementing this into `melos version` is the easiest option as it already has access to the current changelog. There doesn't seem to be any support for getting the changelog of an already created version
- Using GITHUB_TOKEN or CI_JOB_TOKEN (GitLab CI) only works in a CI environment. For people creating releases locally this could either fall back to the simple URL version or melos could integrate with the [GitHub CLI](https://cli.github.com/manual/gh_release_create) and [GitLab CLI](https://glab.readthedocs.io/en/latest/release/create.html)

[This](https://github.com/Almighty-Alpaca/melos/releases/tag/melos-v2.8.0) is an example of an unedited release (except for the obvious `EXAMPLE` in the title) that is currently generated by the first version based on the main branch of my fork. If this is already enough as a first step I'd gladly open a PR for this.

**Comment by blaugold** [comment]
@donheshanthaka This issue is for implementing a feature in Melos that automates the creation of GitHub releases as part of running `melos version`.

@Almighty-Alpaca Thank you for sharing those details!

**Comment by blaugold** [comment]
Fixed by #406
